usr/lib/cca_stdll/cca_specific.c

```diff
@@ -9823,8 +9823,33 @@ static CK_RV cca_handle_apqn_event(STDLL_TokData_t *tokdata,
         __sync_or_and_fetch(&cca_private->inconsistent, TRUE);
         TRACE_ERROR("CCA master key setup is inconsistent, all crypto operations will fail from now on\n");
         OCK_SYSLOG(LOG_ERR, "CCA master key setup is inconsistent, all crypto operations will fail from now on\n");
-    } else {
-        __sync_and_and_fetch(&cca_private->inconsistent, FALSE);
+        return rc;
+    }
+
+    __sync_and_and_fetch(&cca_private->inconsistent, FALSE);
+
+    /* Re-check after APQN set change if protected key support is available */
+    rc = cca_get_min_card_level(tokdata);
+    if (rc != CKR_OK) {
+        TRACE_WARNING("Could not re-determine min card level, protected key support not available.\n");
+        return rc;
+    }
+
+    if (cca_private->min_card_version.ver < 7) {
+        TRACE_WARNING("Min card version now is %d, protected key support not available on this system.\n",
+                   cca_private->min_card_version.ver);
+        /* Disable pkey */
+        __sync_and_and_fetch(&cca_private->pkey_wrap_supported, 0);
+    } else if (cca_private->pkey_wrap_supported == 0 &&
+               !ccatok_pkey_option_disabled(tokdata)) {
+        /* get firmware WKVP, this will enable PKEY on success */
+        rc = ccatok_pkey_get_firmware_wkvp(tokdata);
+        if (rc != CKR_OK) {
+            OCK_SYSLOG(LOG_WARNING,
+                "%s: Warning: Could not get wkvp, protected key support not available.\n",
+                __func__);
+            TRACE_WARNING("Could not get wkvp, protected key support not available.\n");
+        }
```

Comment:
This is a good start, but you still need proper locking when updating the min_card_version in cca_get_min_card_level():

- Add `pthread_rwlock_t min_card_version_rwlock;` to `struct cca_private_data`
- Initialize the lock in `token_specific_init()`: `pthread_rwlock_init(&cca_private->min_card_version_rwlock, NULL);` (plus error checking)
- Free lock in token_specific_final(): `pthread_rwlock_destroy(&cca_private->min_card_version_rwlock);`
- Write-Lock in cca_get_min_card_level() where you copy the version from `card_level_data` to `cca_private->min_card_version`: `pthread_rwlock_wrlock(&cca_private->min_card_version_rwlock);` plus error checking and `pthread_rwlock_unlock(&cca_private->min_card_version_rwlock);` after the copy.  Do not hold the lock while you iterate over the APQNs, just while you are copying the version from `card_level_data` to `cca_private->min_card_version`.
- Read lock in `ccatok_pkey_get_firmware_wkvp() `where you check the min-card_version: `pthread_rwlock_rdlock(&cca_private->min_card_version_rwlock);` plus error checking and `pthread_rwlock_unlock(&cca_private->min_card_version_rwlock);`. I.e. its this line that needs the lock:  `if (cca_data->min_card_version.ver < 7) {`
- Read lock in cca_handle_apqn_event() where you check the min-card_version: `pthread_rwlock_rdlock(&cca_private->min_card_version_rwlock);` plus error checking and `pthread_rwlock_unlock(&cca_private->min_card_version_rwlock);`. I.e. its this line that needs the lock:  `if (cca_data->min_card_version.ver < 7) {`
- in ccatok_pkey_get_firmware_wkvp() use __sync_or_and_fetch() to set pkey_wrap_supported to TRUE.

On your discussion points:
1. do we need to check the CCA lib version also? I think not, because
   the lib cannot be exchanged dynamically.
-> no the lib cannot change.

2. In the previous logic, the routine returned CKR_OK when cca_check_mks
   returned something != CKR_OK. I'm now returning this error rc, which
   seems more appropriate
-> hm no, better not. The event handling function did its work properly, and it flagged the token as inconsistent in that case. So no reason to return an error. Returning an error from the event handler function means that the event handling as such failed, which is not the case. So please change it back to return CKR_OK. Also, I would not skip the CCA version checking in that case either. Let it continue just as it was before.


3. RW lock around updating the min card level and the pkey_wrap_supported
   variable?
-> yes, see above. For updating pkey_wrap_supported __sync_or_and_fetch / __sync_and_and_fetch is sufficient.

Reply:
Regarding the travis errors, I guess I need #ifndef NO_PKEY around the new code block. Will re-push to check ... 
Also, changed the return above the new code block back to 'return CKR_OK'. 
Other changes will follow ...

Reply:
Please move the #ifdef after the call to cca_get_min_card_level(), we do want to update the version info in any case, regardless if pkey is supported on the platform or not.

Reply:
Done.

Reply:
Looks good. Lets merge this with commit "[CCA pkey option: add check for minimum card version](https://github.com/opencryptoki/opencryptoki/pull/676/commits/5c3b4e5f64814b4afaedff1eb3102215656a2647)".